internal/loader/configuration_client_manager.go

```diff
@@ -0,0 +1,437 @@
+		Client:         staticClient,
+		BackOffEndTime: metav1.Time{},
+		FailedAttempts: 0,
+	}}
+
+	return manager, nil
+}
+
+func (manager *ConfigurationClientManager) GetClients() ([]*ConfigurationClientWrapper, error) {
+	currentTime := metav1.Now()
+	clients := make([]*ConfigurationClientWrapper, 0)
+	for _, clientWrapper := range manager.StaticClientWrappers {
+		if currentTime.After(clientWrapper.BackOffEndTime.Time) {
+			clients = append(clients, clientWrapper)
+		}
+	}
+
+	if !manager.ReplicaDiscoveryEnabledEnabled {
+		return clients, nil
+	}
+
+	if currentTime.After(manager.lastFallbackClientAttempt.Time.Add(MinimalClientRefreshInterval)) &&
+		(manager.DynamicClientWrappers == nil ||
+			currentTime.After(manager.lastFallbackClientRefresh.Time.Add(FallbackClientRefreshExpireInterval))) {
+		manager.lastFallbackClientAttempt = currentTime
+		url, _ := url.Parse(manager.endpoint)
+		err := manager.DiscoverFallbackClients(url.Host)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	for _, clientWrapper := range manager.DynamicClientWrappers {
+		if currentTime.After(clientWrapper.BackOffEndTime.Time) {
+			clients = append(clients, clientWrapper)
+		}
+	}
+
+	return clients, nil
+}
+
+func (manager *ConfigurationClientManager) RefreshClients() error {
+	currentTime := metav1.Now()
+	if manager.ReplicaDiscoveryEnabledEnabled &&
+		currentTime.After(manager.lastFallbackClientAttempt.Time.Add(MinimalClientRefreshInterval)) {
+		manager.lastFallbackClientAttempt = currentTime
+		url, _ := url.Parse(manager.endpoint)
+		err := manager.DiscoverFallbackClients(url.Host)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (manager *ConfigurationClientManager) DiscoverFallbackClients(host string) error {
+	srvTargetHosts, err := QuerySrvTargetHost(host)
+	if err != nil {
+		klog.Warningf("Fail to build fall back clients %s", err.Error())
+	}
+
+	// Shuffle the list of SRV target hosts
+	for i := range srvTargetHosts {
+		j := rand.Intn(i + 1)
+		srvTargetHosts[i], srvTargetHosts[j] = srvTargetHosts[j], srvTargetHosts[i]
+	}
+
+	newDynamicClients := make([]*ConfigurationClientWrapper, 0)
+	for _, host := range srvTargetHosts {
+		if isValidEndpoint(host, manager.validDomain) {
+			targetEndpoint := "https://" + host
+			if targetEndpoint == manager.endpoint {
+				continue
+			}
+			client, err := manager.newConfigurationClient(targetEndpoint)
+			if err != nil {
+				return fmt.Errorf("build fallback clients failed, %s", err.Error())
+			}
+			newDynamicClients = append(newDynamicClients, &ConfigurationClientWrapper{
+				Endpoint:       targetEndpoint,
+				Client:         client,
+				BackOffEndTime: metav1.Time{},
+				FailedAttempts: 0,
+			})
+		}
+	}
+
+	manager.DynamicClientWrappers = newDynamicClients
+	manager.lastFallbackClientRefresh = metav1.Now()
+	return nil
+}
+
+func QuerySrvTargetHost(host string) ([]string, error) {
+	results := make([]string, 0)
+
+	_, originRecords, err := net.LookupSRV(Origin, TCP, host)
```

Comment:
We need to use this method `LookupSRV(ctx context.Context, service, proto, name string) (string, []*SRV, error)` passing the context, which means all the calling chain should add context in the parameter